fairscale/nn/data_parallel/fully_sharded_data_parallel.py

```diff
@@ -547,13 +556,24 @@ def summon_full_params(self, recurse: bool = True) -> Generator:
             # Set the state so that we assert when trying to go into
             # forward/backward.
             self.training_state = TrainingState.SUMMON_FULL_PARAMS
-            self._rebuild_full_params()
-            try:
-                yield
-            finally:
-                self._free_full_params()
-                self._use_fp32_param_shard()
-                self.training_state = TrainingState.IDLE
+            self._rebuild_full_params(full_precision=full_precision)
+            with contextlib.ExitStack() as stack:
+                if full_precision and self.mixed_precision and self.flatten_parameters and self.module.is_flattened:
+                    # This is a special case where the full precision params are
+                    # in new FP32 storage, but the flattened views still point
+                    # to the old FP16 storage (_full_param_padded), so we
+                    # unflatten manually with the new FP32 storage.
+                    assert len(self.params) == 1
+                    assert isinstance(self.module, FlattenParamsWrapper)
+                    stack.enter_context(self.module.unflatten_params(recurse=False, flat_param=self.params[0]))
+                try:
+                    yield
+                finally:
+                    stack.close()
+                    if not full_precision:
```

Comment:
```suggestion
                    # _rebuild_full_param() above didn't create full params. So no need to free here.
                    if not full_precision:
```

Maybe another way to do this:

full_param_allocated = _rebuild_full_params()
...

if full_param_allocated:
   _free_full_params()